From 96e276c78927a5731f564ed34f3d32d577b947df Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:13:05 -0400
Subject: [virt] kvm/ppc: book3s/pr - Enable interrupts earlier

Message-id: <1410765214-16377-50-git-send-email-dgibson@redhat.com>
Patchwork-id: 94736
O-Subject: [PATCH 49/78] KVM: PPC: Book3S: PR: Enable interrupts earlier
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Signed-off-by: David Gibson <dgibson@redhat.com>

commit 3d3319b45eea26df56c53aae1a65adf74c8ab12a
Author: Alexander Graf <agraf@suse.de>
Date:   Fri Nov 29 02:32:31 2013 +0100

    KVM: PPC: Book3S: PR: Enable interrupts earlier

    Now that the svcpu sync is interrupt aware we can enable interrupts
    earlier in the exit code path again, moving 32bit and 64bit closer
    together.

    While at it, document the fact that we're always executing the exit
    path with interrupts enabled so that the next person doesn't trap
    over this.

    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_interrupts.S b/arch/powerpc/kvm/book3s_interrupts.S
index b60b511..9729f17 100644
--- a/arch/powerpc/kvm/book3s_interrupts.S
+++ b/arch/powerpc/kvm/book3s_interrupts.S
@@ -148,6 +148,7 @@ after_sprg3_load:
   * R12      = exit handler id
   * R13      = PACA
   * SVCPU.*  = guest *
+  * MSR.EE   = 1
   *
   */
 
@@ -167,11 +168,6 @@ after_sprg3_load:
  nop
 
 #ifdef CONFIG_PPC_BOOK3S_64
- /* Re-enable interrupts */
- ld r3, HSTATE_HOST_MSR(r13)
- ori r3, r3, MSR_EE
- MTMSR_EERI(r3)
-
  /*
   * Reload kernel SPRG3 value.
   * No need to save guest value as usermode can't modify SPRG3.
diff --git a/arch/powerpc/kvm/book3s_rmhandlers.S b/arch/powerpc/kvm/book3s_rmhandlers.S
index c78ffbc..9eec675 100644
--- a/arch/powerpc/kvm/book3s_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_rmhandlers.S
@@ -153,15 +153,11 @@ _GLOBAL(kvmppc_entry_trampoline)
 
  li r6, MSR_IR | MSR_DR
  andc r6, r5, r6 /* Clear DR and IR in MSR value */
-#ifdef CONFIG_PPC_BOOK3S_32
  /*
   * Set EE in HOST_MSR so that it's enabled when we get into our
-  * C exit handler function.  On 64-bit we delay enabling
-  * interrupts until we have finished transferring stuff
-  * to or from the PACA.
+  * C exit handler function.
   */
  ori r5, r5, MSR_EE
-#endif
  mtsrr0 r7
  mtsrr1 r6
  RFI
-- 
1.7.1